Warn about multiple conflicting #[repr] hints

lfairy opened the pull request:
Closes #34622

rust-highfive commented:
r? @pnkfelix

(rust_highfive has picked a reviewer for you, use r? to override)

Aatch commented:
I wonder if this should instead be a warning instead of an error? Also, @nagisa is the use of `#[repr(C, u32)]` in the failing test deliberate?

nagisa commented:
@Aatch what do you mean/are you referring to?

Aatch commented:
@nagisa one of the tests is failing, `mir_adt_construction.rs` because it uses `#[repr(C, u32)]`. You wrote the test, so I pinged you about it.

lfairy commented:
Okay, fixed the error in `mir_adt_construction.rs`. Lets see if Travis likes it now.

nagisa commented:
Ah, no, I don’t think it was intentional.

lfairy commented:
Okay, I've switched to a warning instead.

Should we do a crater run with this patch? I'm curious as to which crates make this mistake in the wild.

lfairy commented:
Just noticed this weird test failure:

```
failures:

---- [pretty] pretty/conflicting-repr-hints.rs stdout ----
	
error: pretty-printed source does not match expected source

expected:
------------------------------------------
// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#![allow(dead_code)]

#[repr(C)]
enum A { A, }

#[repr(u64)]
enum B { B, }

 //~ WARNING conflicting representation hints
#[repr(C, u64)]
enum C { C, }

 //~ WARNING conflicting representation hints
#[repr(u32, u64)]
enum D { D, }

#[repr(C, packed)]
struct E(i32);

fn main() { }

------------------------------------------
actual:
------------------------------------------
// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#![allow(dead_code)]

#[repr(C)]
enum A { A, }

#[repr(u64)]
enum B { B, }

//~ WARNING conflicting representation hints
#[repr(C, u64)]
enum C { C, }

//~ WARNING conflicting representation hints
#[repr(u32, u64)]
enum D { D, }

#[repr(C, packed)]
struct E(i32);

fn main() { }

------------------------------------------
```

Anyone know what's going on here? The test seems to expect an extra space (??)

bors commented:
:umbrella: The latest upstream changes (presumably #35764) made this pull request unmergeable. Please resolve the merge conflicts.

lfairy commented:
I've rewritten the patch to use the HIR machinery. It should be ready to merge now.

Thanks @eddyb for your help on IRC!

eddyb commented on src/librustc/diagnostics.rs:
```diff
@@ -1726,6 +1726,7 @@ fn cookie() -> ! { // error: definition of an unknown language item: `cookie`
 
 register_diagnostics! {
 //  E0006 // merged with E0005
+    E0011, // conflicting representation hints
```
You should pick the first available number *after* the highest in use.

lfairy commented:
@eddyb I've changed the error code to the one after the last.

Should I squash the two commits, or leave them as-is?

eddyb commented:
@lfairy Squashing/amending is preferred in such cases.

lfairy commented:
Sure, I've squashed it now.

bors commented:
:umbrella: The latest upstream changes (presumably #36049) made this pull request unmergeable. Please resolve the merge conflicts.

lfairy commented:
Rebased.

bors commented:
:umbrella: The latest upstream changes (presumably #36066) made this pull request unmergeable. Please resolve the merge conflicts.

lfairy commented:
Rebased again.

By the way: is anyone looking at this PR? It's been open for a while now, and I'd rather get it merged soon before the diagnostics system gets more churn.

Aatch commented:
@bors r+

bors commented:
:pushpin: Commit e2e5807 has been approved by `Aatch`

<!-- @bors r=Aatch e2e580700302172e4306f57c37312c6026958c0a -->

bors commented:
:hourglass: Testing commit e2e5807 with merge b96f323...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2351)

lfairy commented:
Apparently someone else added a diagnostic code in the mean time 😢 Should be fixed now.

@Aatch re-r?

eddyb commented:
@bors r=Aatch

bors commented:
:pushpin: Commit 42b75a5 has been approved by `Aatch`

<!-- @bors r=Aatch 42b75a5c18463b4fef98267663fab13d265eac3e -->

bors commented:
:hourglass: Testing commit 42b75a5 with merge 603d9cc...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10372), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10426), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1660), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1485), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3644), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10366), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10327), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/2299), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/10304), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3526), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5476), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10405), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10378), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2323), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1502), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5492), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2353), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5468), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4660), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1678), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5473), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2349)
Approved by: Aatch
Pushing 603d9ccfbe564e280389e9b3e880d0c53623433d to master...
